### PR TITLE
fireabase-2-que-4-code

### PR DESCRIPTION
**Pull request by ashok-choudhary-0**

### CODE REVIEW

File: controller/userController.js [hunk 1]
```diff
@@ -40,4 +40,52 @@ const updateUserDetails = async (req, res) => {
...
+    const slugUrl = `http://localhost:8000/user/create-new-post/${slugify(bodyData.slug)}`
+    const newPost = await firestore().collection("posts").doc(bodyData.uid).set({
+      title: bodyData.title, description: bodyData.description, slug: slugUrl, updatedAt: createdAt, createdAt, updatedBy: `${user._fieldsProto.firstName.stringValue} ${user._fieldsProto.lastName.stringValue}`
+    })
+    res.status(200).send({ message: "Post created successfully", newPost })
+  } catch (err) {
+    res.status(500).send(err.message)
+  }
+}
+const uploadImage = async (req, res) => {
```
Comment (meanstack9716) [inline controller/userController.js:62]: which Image? It should be uploadUserProfilePhoto like this.

File: controller/userController.js [hunk 2]
```diff
@@ -40,4 +40,52 @@ const updateUserDetails = async (req, res) => {
...
+  try {
+    const fileName = Date.now() + '-' + file.originalname
+    const bucket = admin.storage().bucket();
+    const fileRef = bucket.file(`images/${uid}/${fileName}`);
+    await fileRef.save(file.buffer, {
+      metadata: {
+        contentType: file.mimetype,
+      },
+    });
+    const downloadURL = await fileRef.getSignedUrl({
```
Comment (meanstack9716) [inline controller/userController.js]: ImageDownloadUrl

File: controller/userController.js [hunk 3]
```diff
@@ -40,4 +40,52 @@ const updateUserDetails = async (req, res) => {
...
+  const validateFieldErrors = validationResult(req);
+  if (!validateFieldErrors.isEmpty()) {
+    res.status(404).send({ validateFieldErrors })
+    return;
+  }
+  const bodyData = matchedData(req);
+  try {
+    const user = await firestore().collection("users").doc(bodyData.uid).get();
+    const createdAt = firestore.FieldValue.serverTimestamp()
+    const slugUrl = `http://localhost:8000/user/create-new-post/${slugify(bodyData.slug)}`
```
Comment (meanstack9716) [inline controller/userController.js]: @ashok-choudhary-0  slug implementation is not ok

File: controller/userController.js [hunk 4]
```diff
@@ -40,4 +40,52 @@ const updateUserDetails = async (req, res) => {
...
+    const newPost = await firestore().collection("posts").doc(bodyData.uid).set({
+      title: bodyData.title, description: bodyData.description, slug: slugUrl, updatedAt: createdAt, createdAt, updatedBy: `${user._fieldsProto.firstName.stringValue} ${user._fieldsProto.lastName.stringValue}`
```
Comment (meanstack9716) [inline controller/userController.js]: format the code and updatedBy should be userId @ashok-choudhary-0

File: controller/userController.js [hunk 5]
```diff
@@ -40,4 +40,52 @@ const updateUserDetails = async (req, res) => {
...
+    res.status(200).send({ message: "Post created successfully", newPost })
+  } catch (err) {
+    res.status(500).send(err.message)
+  }
+}
+const uploadImage = async (req, res) => {
+  const { uid } = req.headers;
+  const file = req.file;
+  const validateFieldErrors = validationResult(req);
+  if (!validateFieldErrors.isEmpty()) {
```
Comment (meanstack9716) [inline controller/userController.js]: can't we move this function to a middleware? i am seeing this same function in createPost and uploadImage @ashok-choudhary-0

File: router/userRouter.js [hunk 6]
```diff
@@ -1,7 +1,11 @@
 const router = require("express").Router();
-const { validateFields } = require("../middlewares/userMiddleWare")
+const { validateFields, validatePostFields } = require("../middlewares/userMiddleWare")
 const userController = require("../controller/userController")
+const { uploadImageToFirebase } = require("../controller/uploadImage");
+const { header, body } = require("express-validator");
 router.post("/register", validateFields, userController.userRegister)
 router.post("/create-user-firebase-token", userController.createFirebaseToken)
 router.post("/update-user", userController.updateUserDetails)
+router.post("/create-new-post", validatePostFields, userController.createPost)
```
Comment (meanstack9716) [inline router/userRouter.js]: create new post be in postcontroller, not un user controller @ashok-choudhary-0

**Review by meanstack9716 — CHANGES_REQUESTED** [review]

File: controller/postController.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
+const { matchedData } = require("express-validator");
+const { firestore } = require("firebase-admin");
+const slugify = require('slugify')
+const createPost = async (req, res) => {
+  const { description, slug, title, uid } = matchedData(req);
```
Comment (yuvrajInno) [inline controller/postController.js]: can you create slug from title, do not take i from payload @ashok-choudhary-0

File: router/userRouter.js [hunk 8]
```diff
@@ -1,7 +1,12 @@
...
+const { validateFields, validatePostFields, validateFieldErrors } = require("../middlewares/userMiddleWare")
 const userController = require("../controller/userController")
+const postController = require("../controller/postController")
+const { uploadImageToFirebase } = require("../controller/uploadImage");
+const { header } = require("express-validator");
 router.post("/register", validateFields, userController.userRegister)
 router.post("/create-user-firebase-token", userController.createFirebaseToken)
 router.post("/update-user", userController.updateUserDetails)
+router.post("/create-new-post", [validatePostFields, validateFieldErrors], postController.createPost)
+router.post("/upload-image", [[header("uid").notEmpty()], validateFieldErrors, uploadImageToFirebase.single("image")], userController.uploadImage)
```
Comment (yuvrajInno) [inline router/userRouter.js]: your function does not belongs to a user, its for any image upload create another controller if it, it should not be un user controller @ashok-choudhary-0

**Review by yuvrajInno — CHANGES_REQUESTED** [review]

**Pull request merged**